GPU optimization & Pseudofermions

Chrisgant opened the pull request:
- Improved Hubbard Exp Fermion matrix
- Hubbard Exp dMdPhi for Pseudofermions
- GPU Profiling
- CG overhaul for performance improvements

Marcel-Rodekamp committed:
Share push

Marcel-Rodekamp committed:
Improved device accessibility: Removing a const and putting in a repr_ from torch device

Marcel-Rodekamp committed:
Added a tensor factory eye

Marcel-Rodekamp committed:
Bug fix in Tensor::shift

Marcel-Rodekamp committed:
Added NSL::LinAlg::inner

Marcel-Rodekamp committed:
Added logging to the thermalization in HMC

Marcel-Rodekamp committed:
Added functionality to populate for a source vector that is actually a source matrix into hubbardExp

Marcel-Rodekamp committed:
Merged the correlator example files

Marcel-Rodekamp committed:
Added optional parameters to example_MCMC

Marcel-Rodekamp committed:
Cleaned up h5_io code, added copy savety for GPU tensors

Marcel-Rodekamp committed:
Added two point correlation function

Marcel-Rodekamp committed:
Updated example parameter file

Commit:
modifications to tangent plane stuff

Commit:
added h5 flush call

Commit:
fixed indexing of configurations in h5_io

Commit:
I have no idea what I'm doing!

Merge branch 'feature/Measurements' of https://github.com/Marcel-Rodekamp/NSL into feature/Measurements

Commit:
fixed example_MCMC.cpp so that we can run with a constant shift

Commit:
NSL can run with an offset now

Commit:
changed 'tangentPlane'--->'offset' to keep things more general

Marcel-Rodekamp committed:
Added a overwrite mechanic

Marcel-Rodekamp committed:
Removed falsy feature for expanded fermion matrix application

Marcel-Rodekamp committed:
Inserted an explicit loop over columns in correlators; multiple source calculation wasn't implemented correctly

Marcel-Rodekamp committed:
Bug fix in Tensor::expand

Marcel-Rodekamp committed:
Added LinAlg::tensordot

Marcel-Rodekamp committed:
Added new index objects: Ellipsis (like pythons '[:]') and NewDim (like pythons '[None]')

Marcel-Rodekamp committed:
Added a parameter pack expansion tool to check if a pack contains derived objects

Marcel-Rodekamp committed:
Added operator== to the device object

Marcel-Rodekamp committed:
Added isDerived concept and expanded isIndexer woth NSL::Indexer with it

Marcel-Rodekamp committed:
Simplified return type deduction for complex imag and real

Marcel-Rodekamp committed:
Added defined propery query to Tensor

Marcel-Rodekamp committed:
Included Tensor/Factory/random.tpp into tensor.hpp

Marcel-Rodekamp committed:
Added Tensor::log

Marcel-Rodekamp committed:
but fix with boundary shift

Marcel-Rodekamp committed:
Added tensor reduction mean, var

Marcel-Rodekamp committed:
copied real and imag into Tensor/Impl/realImag.tpp for better accessibility

Marcel-Rodekamp committed:
Reworked the whole random Access and simplified accessing GPU Tensors now

Marcel-Rodekamp committed:
Added Tensor::rand(low,high) and Tensor::randn(mean,std)

Marcel-Rodekamp committed:
Added Tensor::device() query

Marcel-Rodekamp committed:
Added randn_like factories

Marcel-Rodekamp committed:
Prepared the markovState for pseudofermions

Marcel-Rodekamp committed:
Prepare HMC for pseudo fermions

Marcel-Rodekamp committed:
Implemented Configuration elementwise multiplication

Marcel-Rodekamp committed:
Made CG reporting Debug (thus just clutters files for pseudofermions)

Marcel-Rodekamp committed:
Added LinAlg::log

Marcel-Rodekamp committed:
Added LinAlg::exp(Configuration)

Marcel-Rodekamp committed:
Added LinAlg::tensordot

Marcel-Rodekamp committed:
added H5IO::read/write(scalar)

Marcel-Rodekamp committed:
Added a Pseudofermion Action (Hubbard Fermion matrices are not ready for this yet)

Marcel-Rodekamp committed:
Added NSL::Random::choice

Marcel-Rodekamp committed:
Implemented a Suit for U1 including a WilsonGaugeAction; U1Leapfrog; WilsonLoop Measurement

Marcel-Rodekamp committed:
Implemented a Wilson Fermion Matrix for the U1 gauge theory

Marcel-Rodekamp committed:
Handled underflow edge case for getMatchingDigits in tests.

Marcel-Rodekamp committed:
Bug fix in doxygen

Marcel-Rodekamp committed:
Forgot one CMakeLists.txt

Commit:
first commit of work related to momentum correlators

Commit:
working version of momentum correlators

petio9671 committed:
Implemented two-body correlators. Its usage is
almost similar to the TwoPointCorrelator but a bit expanded.

petio9671 committed:
Cleaned up some unnecessary variables and comments.
Now the code is a bit easier to read.

petio9671 committed:
Fixed indexing issue of the correlator matrix.
Commented out the kronekerProduct function.

petio9671 committed:
Implemented NSL::LinAlg::flip function which has a python equivalent torch.flip. It flips the order along a given axis or axes

petio9671 committed:
Implemented state I=1 S=1 Iz=0 Sz=1

petio9671 committed:
Implemented the whole connected I=1, S=1 table

petio9671 committed:
Implemented all channels with connected diagrams from the Spin-Isospin tables

petio9671 committed:
Changed the default CG criterion to 1e-12 and the output to be error not error^2

petio9671 committed:
Implemented sqrt function

petio9671 committed:
All connected diagram channels are working

petio9671 committed:
Fixed indexing and sign issues

petio9671 committed:
Changed the names of the indices in the channels so that it is easier to read

petio9671 committed:
The overlap flag is now working correctly when we want to generate configurations i.e. using example_MCMC

Commit:
fixed checkpointing issues with example_k_correlators.cpp and twoBodyCorrelationFunction.tpp.  Also changed some things in twoBodyCorrelationFunction.tpp so that it compiles on more mac machines

petio9671 committed:
Fixed an issue which shifted the Non-Interacting correlators by an amount of 'Number of Timesources'

Marcel-Rodekamp committed:
Schare Push

Marcel-Rodekamp committed:
Merge branch 'devel' into feature/SchwingerModel

petio9671 committed:
Fixed an indexing issue in channel I1S1Iz1Sz0

petio9671 committed:
Now we write out the momenta at which we measure to h5/Momenta

petio9671 committed:
Fixed another indexing issue (I1S1Iz1Sz0)

petio9671 committed:
Added vector of vectors for reading momenta

petio9671 committed:
Forgot to change precision in 2 Constructors -- Fixed

petio9671 committed:
Added checkpointing to example_MCMC, example_k_correlators, and example_2_correlators

Marcel-Rodekamp committed:
Removed some non compiling left overs

petio9671 committed:
Updated the documentataion/comments of the 2-body

petio9671 committed:
Now inner product can take a dim to reduce along.Also expand can take a dim to expand on.

petio9671 committed:
Two-Body correlator supports GPU

petio9671 committed:
Fixed expand(newSize, newDim) function

petio9671 committed:
One- and Two-body correlator executables support GPU

Chrisgant committed:
Implemented Hubbard Pseudofermions

Chrisgant committed:
Merge branch 'feature/PseudoFermions' of https://github.com/Marcel-Rodekamp/NSL into feature/PseudoFermions

petio9671 committed:
Fixed S_z = 0. Now it produces the correct output and there is S degeneracy as expected.

Chrisgant committed:
Gattringer sign fix, Pseudo-Leapfrog converging

Chrisgant committed:
HMC executable Hubbardmodel with Pseudofermions

Chrisgant committed:
add hidden directories in base dir to gitignore

Chrisgant committed:
Merge Correlator branch into Pseudofermions for testing

Chrisgant committed:
created example four site yaml

Chrisgant committed:
copies all yamls in example directory

petio9671 committed:
Changed conj() functions for NSL::LinAlg::conj() because it does not mutate the data

petio9671 committed:
Added an example of broadcasted multiplication

Chrisgant committed:
switch back to 2 times real part in Pseudofermion action

Chrisgant committed:
create executable for Pseudofermion Hubbard HMC

Chrisgant committed:
update example params for momentum correlators

Chrisgant committed:
Basic python analysis scripts

petio9671 committed:
Now the running average is moving

petio9671 committed:
Merge branch 'feature/acceptanceRate' into feature/twoBodyCorrelators

petio9671 committed:
We do not need to read the acceptance rate. We can do it in post process. The running acceptance rate is calculated with a window of a 100. When we checkpoint it should monotonically increase until the 100th configuration of the run is reached and then fluctuate.

petio9671 committed:
Combined functions removeData and trimData into function deleteData. deleteData deletes the node that is given if it exists. The overwrite flag should be defined by the user.

petio9671 committed:
Added Multi RHS to Two-Body measurements

Chrisgant committed:
speed up hubbardExp functions

Chrisgant committed:
using previous inverse as starting guess for CG

Chrisgant committed:
Merge branch 'feature/PseudoFermions' of https://github.com/Marcel-Rodekamp/NSL into feature/PseudoFermions

Chrisgant committed:
hotfix

Chrisgant committed:
Tom's correlator analysis notebook

Chrisgant committed:
setup space for comparing implementations

Chrisgant committed:
print CG convergence only in debug mode

petio9671 committed:
This is the first commit on this branch.
All of the modifications to the files are done to implement an
autotuner.

Added an additional templated function to read Type objects from a
h5file. And show a coresponding example.

Now there is not a whole thermalization sequence when generating a new
Markov Chain. This functionality is "outsourced" to a python autotuner
which calls the "thermalizaiton" step one at a time.

Added thermalFlag parameter to show if a file is thermalized or not.
There is no need for Nmd and Ntherm in yaml file anymore. These two are
determined by the autotuner.

Chrisgant committed:
reverting changes to FermionMatrix

Chrisgant committed:
improved benchmark for fermion matrix comparison

Chrisgant committed:
dMdaggerdPhi & accepting new implementations to replace old ones

Chrisgant committed:
Cleanup of FermionMatrix development

Chrisgant committed:
Bugfix: imaginary part of start phi set to constant

Chrisgant committed:
removed leftover from CmakeList

Chrisgant committed:
tuned example parameter file

Chrisgant committed:
Notebook to generate exact correlators

Chrisgant committed:
Merge branch 'feature/PseudoFermions' of https://github.com/Marcel-Rodekamp/NSL into feature/PseudoFermions

Chrisgant committed:
minor update to notebooks

petio9671 committed:
Implemented 2 stage autotuning.

Added two different flags signaling in which stage of the tuning
we are.
Now meta data contains the right values of the flags.

petio9671 committed:
Fixed the acceptance probability output of rejected configurations

Chrisgant committed:
initial guess option for CG

Chrisgant committed:
added RK4 params to example yaml

Chrisgant committed:
stepwise flow executable

Chrisgant committed:
include NVTX for profiling

Chrisgant committed:
NVTX ranges for HMC

Chrisgant committed:
limit prfiling range

Chrisgant committed:
NVTX ranges leapfrogRealForce

Chrisgant committed:
NVTX ranges PseudoHMC

Chrisgant committed:
remove profiling.hpp from NSL.hpp to avoid redefinition

Chrisgant committed:
NVTX ranges PseudoFermionAction

Chrisgant committed:
Profiling and Cuda Graphs for CG (unclean)

Chrisgant committed:
fixed temporary nvtx path

Chrisgant committed:
Cuda Graphs for CG with initial guess

petio9671 committed:
Autotuning of MCMC is done in a new example file called example_autotune_MCMC.cpp

Chrisgant committed:
CG with CUDA-graphs (rough) cleanup

petio9671 committed:
This is a python wrapper of the NSL MCMC that performs autotuning.
It does a 2 stage autotuning with working checkpointing.

The code is taken from CNS and it is adjusted to work with NSL.

Chrisgant committed:
fixed missing Nt

Chrisgant committed:
Merge remote-tracking branch 'origin/devel' into feature/GPU_optimization

Chrisgant committed:
Merge remote-tracking branch 'origin/devel' into feature/GPU_optimization

Chrisgant committed:
Merge branch 'feature/GPU_optimization' of https://github.com/Marcel-Rodekamp/NSL into feature/GPU_optimization

Chrisgant committed:
Merge branch 'feature/GPU_optimization' of https://github.com/Marcel-Rodekamp/NSL into feature/GPU_optimization

Chrisgant committed:
Merge branch 'feature/GPU_optimization' of https://github.com/Marcel-Rodekamp/NSL into feature/GPU_optimization

petio9671 committed:
Fixed graph directory and a minor bug

Chrisgant committed:
Merge with autotuner

Chrisgant committed:
including nvtx properly with CMake

petio9671 committed:
Changed the graph filename to (h5 file name)-graph_counter.png

petio9671 committed:
Now the autotuner will continue into production seamlessly

Chrisgant committed:
Cuda option at compile time

Chrisgant committed:
preventing pointer issues with std::shared_ptr

petio9671 committed:
Added chain-batch scripts for generating simulations based on CNS chain-batch scripts

petio9671 committed:
Merge branch 'devel' into feature/autotuner

Chrisgant committed:
include missing measurement

Chrisgant committed:
major overhaul of solvers using more reliable shared pointers

Chrisgant committed:
added USE_CUDA flag, to compile GPU support on cpu only machine (e.g. login node)

Chrisgant committed:
Fixed GPU optimization (preliminary! only works with --GPU right now)

Chrisgant committed:
improved device info

Chrisgant committed:
CG cleanup, running optimization on first function call and skipping if device is cpu

Chrisgant committed:
Merge remote-tracking branch 'origin/feature/autotuner' into feature/GPU_optimization

Chrisgant committed:
const bugfix

Chrisgant committed:
fixed merge mistake

Chrisgant committed:
Changed autotuner to Pseudofermions

Chrisgant committed:
activate GPU capabilities of autotuner

petio9671 committed:
Autotuning is done on CPUs and production on GPUs

Chrisgant committed:
fixed style and typos

Chrisgant committed:
fix sideeffect of batching by limiting CG convergence

Chrisgant committed:
merge Autotuner

Chrisgant committed:
device related optimization

Chrisgant committed:
fix normal distribution

Chrisgant committed:
improve structure and documentation

Chrisgant committed:
jupyter notebook for exact correlators

Chrisgant committed:
merge devel

Chrisgant committed:
If torch is not found, try to query pip for it's location

Chrisgant committed:
introduced new flags in readme

luutom commented:
@Chrisgant So there are conflicting files.  Before I start the arduous task of going over your commits, can you resolve this first?

Chrisgant committed:
Merge branch 'devel' into feature/GPU_optimization

luutom reviewed (changes requested):
We don't want this going into devel, since there are 'hard-wired' directories

luutom committed:
Update example_Pseudofermion.cpp

Added relevant comments to the code to describe commented (but useful) snippets

luutom committed:
Update example_param.yml

Remove redundant entry of Nt (this was Marcel's fault)

luutom committed:
Delete Executables/Hubbard/example_param.yml

Deleted redundant yml file

luutom committed:
Delete Scripts/visualize_flow.ipynb

Unfinished file, so it was deleted

Chrisgant committed:
Fixed issue where torch_cuda_version was not set

Chrisgant committed:
copying yaml files automatically during build

Chrisgant committed:
Merge remote-tracking branch 'origin/devel' into feature/GPU_optimization

luutom committed:
Update hubbardExp.tpp

Added comments to explain what the hell is going on.

luutom commented on src/NSL/Tensor/Impl/factory.tpp:
I don't think we should merge this change

luutom committed:
Update factory.tpp

Making it consistent with devel branch

luutom reviewed (approved):
God help us!

Commit:
resolving conflicts between Christoph and Petar.  It is hard work!

Commit:
more template to shit from U1 stuff

Commit:
Now including Petar's fix to the timesource averaging

Merge branch 'devel' into feature/GPU_optimization
